include/yui/types.h

```diff
@@ -0,0 +1,142 @@
+#ifndef TYPES_H
+#define TYPES_H
+
+#include <imgui.h>
+
+#include <compare>
+
+namespace yui {
+
+/**
+ * @brief   Internal vec2 type
+ */
+struct vec2 {
+    constexpr vec2() = default;
+    constexpr vec2(float x, float y):
+        data{ x, y }
+    {}
+    constexpr vec2(ImVec2 obj):
+        data{ obj.x, obj.y }
+    {}
+
+    auto operator<=>(const vec2&) const = default;
+
+    inline vec2& operator+=(const vec2& obj) noexcept {
+        data[0] += obj.data[0];
+        data[1] += obj.data[1];
+        return *this;
+    }
+
+    /**
+     * @brief   Division by a scalar
+     */
+    inline vec2& operator/=(float scalar) noexcept {
+        data[0] /= scalar;
+        data[1] /= scalar;
+        return *this;
+    }
+
+    operator ImVec2() const {
+        return { data[0], data[1] };
+    }
+
+    float& x() { return data[0]; }
+    float& y() { return data[1]; }
+
+    float x() const { return data[0]; }
+    float y() const { return data[1]; }
+
+    float data[2];
+};
+
+inline vec2 operator+(const vec2& lhs, const vec2& rhs) noexcept {
+    return {
+        lhs.data[0] + rhs.data[0],
+        lhs.data[1] + rhs.data[1]
+    };
+}
+
+inline vec2 operator+(const ImVec2& lhs, const vec2& rhs) noexcept {
+    return {
+        lhs.x + rhs.data[0],
+        lhs.y + rhs.data[1]
+    };
+}
+
+inline vec2 operator+(const vec2& lhs, const ImVec2& rhs) noexcept {
+    return rhs + lhs;
+}
+
+inline vec2 operator-(const vec2& lhs, const vec2& rhs) noexcept {
+    return {
+        lhs.data[0] - rhs.data[0],
+        lhs.data[1] - rhs.data[1]
+    };
+}
+
+inline vec2 operator-(const ImVec2& lhs, const vec2& rhs) noexcept {
+    return {
+        lhs.x - rhs.data[0],
+        lhs.y - rhs.data[1]
+    };
+}
+
+inline vec2 operator-(const vec2& lhs, const ImVec2& rhs) noexcept {
+    return rhs - lhs;
+}
+
+/**
+ * @brief:   Element-wise multiplication
+ */
+inline vec2 operator*(const vec2& lhs, const vec2& rhs) noexcept {
+    return {
+        lhs.data[0] * rhs.data[0],
+        lhs.data[1] * rhs.data[1]
+    };
+}
+
+/**
+ * @brief   Element-wise division
+ */
+inline vec2 operator/(const vec2& lhs, const vec2& rhs) noexcept {
+    return {
+        lhs.data[0] / rhs.data[0],
+        lhs.data[1] / rhs.data[1]
+    };
+}
+
+/**
+ * @brief   Division by a scalar
+ */
+inline vec2 operator/(const vec2& lhs, float rhs) noexcept {
+    return {
+        lhs.data[0] / rhs,
+        lhs.data[1] / rhs
+    };
+}
+
+/**
+ * @brief   Internal vec4 type
+ */
+struct vec4 {
```

Comment:
I still don't know if it is a good idea to mix concepts. While a color can be represented with a `vec4`, it should not be convertible to a mathematical vector. Or is it?

Reply:
Will be removed.